Api revision

espy opened the pull request:
I'm sorry, this is massive. Does the following:
- rewrote API docs and API intro
- restyled all tables and added correct syntax highlighting everywhere
- removed the server API page from the nav
- added automatically inserted "back-to-top"-links
- added smooth scrolling for internal anchors
- added some new styles to improve readability/scannability
- general re-editing of previously updated guides

**Note**: there are three things at the end of `docs.css` that we'll want in the global stylesheet at some point:
- fixes a bug with the invisible sub-header nav being clickable
- improves legibility of links (made them thicker)
- fixed some display problems with the link underlining breaking at smaller font sizes and in different browsers:

Firefox OS X before:
![bildschirmfoto 2015-07-09 um 15 12 19](https://cloud.githubusercontent.com/assets/391124/8595996/f8028d50-264c-11e5-9333-26e1a30013b8.png)

Firefox OS X after:
![bildschirmfoto 2015-07-09 um 15 12 29](https://cloud.githubusercontent.com/assets/391124/8595999/01a955d2-264d-11e5-90a8-a47a99932cab.png)

That's it!
Woop!

espy committed:
Added table styles to docs

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm

espy committed:
Removed unused doc from english API menu

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm

espy committed:
Updated API docs intro text

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm

espy committed:
wip: rewriting hoodie client side API docs

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm

espy committed:
Better anchor headlines

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm

espy committed:
Added smooth scrolling for anchor links

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm

espy committed:
Completed revision of hoodie base API guide

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm

espy committed:
Changed doc lists to use discs instead of circles

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm

espy committed:
API base guide: Added link to itself on GH

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm

espy committed:
Made font in a tags thicker, for better readability

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm

espy committed:
Made font in a tags thicker, for better readability

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm

espy committed:
API base guide: Added link to itself on GH

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm

espy committed:
Merge branch 'api-revision' of github.com:hoodiehq/documentation into api-revision

espy committed:
rewrote accounts api guide

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm

espy committed:
Text formatting in account doc

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm

espy committed:
Added automatic insertion of back-to-top links

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm

espy committed:
Better api doc section headlines

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm

espy committed:
Various improvements to several docs

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm

espy committed:
Updated hoodie.store API docs

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm

espy committed:
Isolated fix for invisible header still being clickable

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm

espy committed:
Revision of concepts page, closes #175

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm

espy committed:
Various smaller corrections throughout

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm

espy committed:
Fixed link underlines breaking at different font sizes/browsers

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm

gr2m commented on en/index.md:
````diff
@@ -18,7 +18,16 @@ All of Hoodie is accessible through a simple script include, just like jQuery or
+```javascript
+// In your front-end code:
+hoodie.account.signUp(username, password);
+```
````
very nice idea to show some code right away!

gr2m commented on en/techdocs/api/client/hoodie.account.md:
```diff
@@ -3,27 +3,25 @@ layout: layout-api
+**source:** 		*[hoodie.js/src/hoodie/account](https://github.com/hoodiehq/hoodie.js/tree/master/src/hoodie/account)*
 
-**<br />after reading this you will know:**
+The account object in the client-side Hoodie API covers all user and authentication-related tasks, and enables you to do previously complex operations, such as signing up a new user, with only a few lines of frontend code. Since [data in Hoodie is generally bound to a user](/en/hoodieverse/glossary.html#private-user-store), it makes sense to familiarise yourself with **account** before you move o ...
```
I wonder if we can replaces "authentication-related tasks" with something else, just because "tasks" have a meaning in Hoodie land

espy replied:
Changed to "operations"

gr2m commented on en/techdocs/api/client/hoodie.account.md:
```diff
@@ -157,18 +156,25 @@ synchronized to.
-<br />
+#### NOTES
 
+All Hoodie data that existed in the browser's local store before signin will be cleared on signin. To prevent data loss, you can pass **{moveData: true}** as the options argument, this will move current data (created anonymously or by another account) to the account the user signs in to.
```
> or by another account

Did you test if that's the case? I'm not sure myself to be honest, if it's possible to sign in to a different account if you are currently signed in

espy replied:
Good catch. It _does_ kinda work, but Hoodie fails to clear the local store before doing the second sign in. So when you sign in to account B while signed into A, you'll get both A and B's todos. If you sign back into A, you'll get A's todos twice, and B's todos once. Sign back into B, you'll have both twice. 

So this should definitely be handled differently, but I'll remove it from the docs for now.

Just to make sure you see this, here's a mention: @gr2m

gr2m commented on en/techdocs/api/client/hoodie.md:
```diff
@@ -8,13 +8,15 @@ locales: en
 ## Introduction
 
-This document describes the functionality of the hoodie base object.
+This document describes the functionality of the hoodie base object. It provides a number of helpers dealing with event handling and connectivity, as well as a unique id generator and a means to set the endpoint which Hoodie communicates with.
```
> helpers

Maybe "helper methods"? But you know better, just a thought

espy replied:
Added :)

gr2m commented on en/techdocs/api/client/hoodie.md:
```diff
@@ -29,253 +31,337 @@ This document describes the functionality of the hoodie base object.
-in the browser's local store. When signing in to an account,
-**hoodie.id()** will return the id of the new user.
-On signout, hoodie.id() gets reset.
+Each Hoodie user is assigned a random, unique, persistent ID on signup. While the user is logged in, **id()** will always return that same ID. When no user is signed in, it will return a different, non-persistent ID for the current anonymous user. Every time a user signs out, the ID for the anonymous user changes.
```
>  Each Hoodie user is assigned a random ... ID on signup

This is not correct, the id is assigned the first time you open a hoodie app, and it gets persisted right away. So on page reload, it remains the same. When you sign up for an account, the Hoodie ID remains the same

espy replied:
I'd understood it that way but wrote it down in an unclear way. Changed to:

> Each user is assigned a random, unique ID when they first visit your Hoodie app, even before signup. They are, in this state, **anonymous users**. Their ID and their data will only be persisted locally. 
> 
> When they sign up, their ID becomes permanently bound to their user account. After every single time they sign out again, **hoodie.id()** returns a new, unique value in their client, because they are now de facto a new, anonymous user again.

gr2m commented on en/techdocs/api/client/hoodie.md:
```diff
@@ -29,253 +31,337 @@ This document describes the functionality of the hoodie base object.
 **version:**      *> 0.2.0*
 
-*Bind an event handler to a certain event so you get informed in case it gets triggered.*
+**Binds a persistent event handler for specified events**
```
what is a "persistent" event handler?

espy replied:
Good point, changed to

> Adds an event handler for the specified events

gr2m commented on en/techdocs/api/client/hoodie.md:
```diff
@@ -29,253 +31,337 @@ This document describes the functionality of the hoodie base object.
+| eventHandler | Function | Function handling triggered event  | yes      |
 
-Hoodie informs you about several things happening. For instance every time a store object has been added, saved or removed. In order to catch those messages you can register a function for handling those events. Those functions are called **eventHandlers**. You can register also for events you created and triggered yourself. See [trigger](#trigger) for further information.
+Hoodie informs you about a lot of internal events, such as data being added, removed or changed, or a user signing in or out. These can be listened to with **on()**. You can also register listeners for events you created and triggered yourself. See [trigger](#trigger) more on this.
```
> See [trigger](#trigger) more on this

See [trigger](#trigger) _for_ more on this?

espy replied:
Corrected!

gr2m commented on en/techdocs/api/client/hoodie.md:
````diff
@@ -29,253 +31,337 @@ This document describes the functionality of the hoodie base object.
+object-type:object-id:event-type
+```
+
+**Object-type** and **event-type** are mandatory. Omitting the **object-id** listens to events from all objects of a type. If you *do* specify an object id, you'll only receive events from that one specific object. If you want to listen for more than one event type at once, use the **change** type, which fires on *add*, *update* and *remove*.
````
> +**Object-type** and **event-type** are mandatory

That's not correct. **Object-type** is optional. You can do `hoodie.store.on("update", handle)`.

espy replied:
Corrected to

> The **event-type** is mandatory. If you want to listen for more than one event type at once, use the **change** type, which fires on _add_, _update_ and _remove_.
> 
> **object-type** and **object-id** are optional: Omitting **object-type** listens to all objects in the store, specifying it listens only to objects of that type. Omitting the **object-id** listens to events from all objects of a type. If you _do_ specify an object id, you'll only receive events from that one specific object.

gr2m commented on en/techdocs/api/client/hoodie.md:
```diff
@@ -29,253 +31,337 @@ This document describes the functionality of the hoodie base object.
+var todoStore = hoodie.store('todo');
+todoStore.on('change, function(changedTodoObject) {
+  console.log('A todo has been changed => ', changedTodoObject);
+});
```
```
todoStore.on('change, function
                    ^ quote missing here
```

espy replied:
:white_check_mark:

gr2m commented on en/techdocs/api/client/hoodie.md:
````diff
@@ -29,253 +31,337 @@ This document describes the functionality of the hoodie base object.
+
+```javascript
+hoodie.account.on('signin, function(username) {
+  console.log('Hello there, '+ username);
````
quote missing after `'signin`

espy replied:
:white_check_mark:

gr2m commented on en/techdocs/api/client/hoodie.md:
```diff
@@ -283,22 +369,37 @@ chat.findAll('message')
+
+We also have an <a href="http://faq.hood.ie" target="_blank">FAQ</a> that could prove useful if things go wrong.
+
+If you find this guide in error or out of date, you could also <a href="https://github.com/hoodiehq/documentation/issues" target="_blank">open an issue</a> or <a href="https://github.com/hoodiehq/documentation/pulls" target="_blank">submit a pull request</a> with your corrections to [this file](https://github.com/hoodiehq/documentation/blob/gh-pages/en/techdocs/api/client/hoodie.md).
```
I'm not sure if linking to `https://github.com/hoodiehq/documentation/pulls` is helpful here. I think I'd leave it out

espy replied:
Good point really, where else would they submit the pull request, they'll got the repo from the next link. I'll remove all instances of this.

gr2m commented on en/techdocs/api/client/hoodie.store.md:
```diff
@@ -210,152 +236,178 @@ hoodie.store('todo')
+      // set dueDate to right now
+      todo.dueDate = Date.now();
+    }
+    return todo;
```
return not needed anymore, please leave it out

espy replied:
:white_check_mark:

gr2m commented on en/techdocs/api/client/hoodie.store.md:
````diff
@@ -363,141 +415,184 @@ Returns undefined.
 hoodie.store.findAll('todo')
   .done(function(allTodos) { /* ... */ });
-</code></pre>
+```
````
this won't work, `hoodie.store.add` is an asynchronous function. It should be

``` js
hoodie.store.add('todo', { title: 'Getting Coffee' })
.done(function () {
  hoodie.store.findAll('todo')
  .done(function(allTodos) { /* ... */ })
})
```

espy replied:
Oof, thanks for catching that. I should have made it clearer that `add` and `findAll` don't actually follow each other in the code, only in time. Your solution is a lot safer :)

gr2m commented on en/techdocs/api/client/hoodie.store.md:
```diff
@@ -363,141 +415,184 @@ Returns undefined.
 
 todoStore.add({ title: 'Getting Coffee' });
 todoStore.findAll()
   .done(function(allTodos) { /*...*/ });
```
see comment above. `todoStore.findAll` would be called before object gets added

espy replied:
:white_check_mark:

gr2m commented on en/techdocs/api/client/hoodie.store.md:
```diff
@@ -363,141 +415,184 @@ Returns undefined.
+
+The **clear** event is triggered when a user signed out, or called **hoodie.account.destroy()**. It is also triggered when the user signs in, to clear any existing local objects before loading the objects from the account the user signed into.
+
+Note that no **remove** events is triggered when the store gets cleared, as the objects do not necessarly get removed for the user's account, but only from the local cache.
```
> get removed for the user's account, but only from the local cache

get removed _from_ ... ?

espy replied:
:white_check_mark:

gr2m commented:
Done :+1:

espy committed:
Added changes from Gregor's review

* * *

This commit was sponsored by The Hoodie Firm.
You can hire The Hoodie Firm:
http://go.hood.ie/thehoodiefirm
